### ISSUE TITLE
LSF resource manager for Lassen@LLNL

### ISSUE THREAD

**Issue by mtitov** [radical-cybertools/radical.pilot#2136]
RP at Lassen finished with the following error:
```
$ cat  agent.0.out
bootstrap agent agent.0
failed agent.0
--------------
RADICAL Utils -- Stacktrace [17628] [MainThread]

titov1    17628  16124  6 21:48 ?        00:00:00                      \_ rp.agent.0
Traceback (most recent call last):
File "/p/gpfs1/titov1/.miniconda3/envs/ve.rp/bin/radical-pilot-agent", line 50, in bootstrap_3
if aid == 'agent.0': agent = rp.Agent_0(cfg, session)
File "/p/gpfs1/titov1/.miniconda3/envs/ve.rp/lib/python3.6/site-packages/radical/pilot/agent/agent_0.py", line 67, in __init__
self._configure_rm()
File "/p/gpfs1/titov1/.miniconda3/envs/ve.rp/lib/python3.6/site-packages/radical/pilot/agent/agent_0.py", line 137, in _configure_rm
cfg=self._cfg, session=self._session)
File "/p/gpfs1/titov1/.miniconda3/envs/ve.rp/lib/python3.6/site-packages/radical/pilot/agent/resource_manager/base.py", line 247, in create
return impl(cfg, session)
File "/p/gpfs1/titov1/.miniconda3/envs/ve.rp/lib/python3.6/site-packages/radical/pilot/agent/resource_manager/lsf.py", line 21, in __init__
ResourceManager.__init__(self, cfg, session)
File "/p/gpfs1/titov1/.miniconda3/envs/ve.rp/lib/python3.6/site-packages/radical/pilot/agent/resource_manager/base.py", line 173, in __init__
% (str(cores_avail), str(self.requested_cores)))
ValueError: Not enough cores available (2 < 44).

--------------
finalize agent.0
```

Resource config:
```
{
    "lassen": {
        "description"                 : "",
        "notes"                       : "",
        "schemas"                     : ["local"],
        "local"                       : 
        {
            "job_manager_hop"         : "fork://localhost/",
            "job_manager_endpoint"    : "lsf://localhost/",
            "filesystem_endpoint"     : "file://localhost/"
        },
        "cores_per_node"              : 44,
        "gpus_per_node"               : 4,
        "default_queue"               : "pbatch",
        "resource_manager"            : "LSF",
        "agent_scheduler"             : "CONTINUOUS",
        "agent_spawner"               : "POPEN",
        "agent_launch_method"         : "JSRUN",
        "task_launch_method"          : "JSRUN",
        "mpi_launch_method"           : "JSRUN",
        "pre_bootstrap_0"             : [
            "module load jsrun",
            "source /p/gpfs1/$USER/.miniconda3/etc/profile.d/conda.sh"],
        "default_remote_workdir"      : "/p/gpfs1/$USER",
        "valid_roots"                 : ["/p/gpfs1/$USER",
                                         "$HOME"],
        "python_dist"                 : "anaconda",
        "virtenv"                     : "ve.rp",
        "virtenv_mode"                : "use",
        "rp_version"                  : "installed"
    }
}
```

Key env variables:
```
export LSB_EFFECTIVE_RSRCREQ="1*{select[(LN) && (type == any)] order[r15s:pg] span[hosts=1]}+40*{select[(CN) && (type == any)] order[r15s:pg] span[ptile=40]}"
export LSB_HOSTS="lassen710 lassen827 lassen827 lassen827 lassen827 lassen827 lassen827 lassen827 lassen827 lassen827 lassen827 lassen827 lassen827 lassen827 lassen827 lassen827 lassen827 lassen827 lassen827 lassen827 lassen827 lassen827 lassen827 lassen827 lassen827 lassen827 lassen827 lassen827 lassen827 lassen827 lassen827 lassen827 lassen827 lassen827 lassen827 lassen827 lassen827 lassen827 lassen827 lassen827 lassen827"
export LSB_MCPU_HOSTS="lassen710 1 lassen827 40 "
export LSB_QUEUE="pbatch"
export LSB_SUB_HOST="lassen709"
export LSB_SUB_RES_REQ="1*{select[LN]span[hosts=1]} + 40*{select[CN]span[ptile=40]}"
```
```
pilot.0000/rp.session.lassen709.titov1.018381.0000.log:1588136139.041 : rp.session.lassen709.titov1.018381.0000 : 42293 : 35184372370416 : INFO     : Found LSB_DJOB_HOSTFILE /g/g90/titov1/.lsbatch/1588136102.1041317.hostfile. Expanded to: ['lassen710', 'lassen827', 'lassen827', 'lassen827', 'lassen827', 'lassen827', 'lassen827', 'lassen827', 'lassen827', 'lassen827', 'lassen827', 'lassen827', 'lassen827', 'lassen827', 'lassen827', 'lassen827', 'lassen827', 'lassen827', 'lassen827', 'lassen827', 'lassen827', 'lassen827', 'lassen827', 'lassen827', 'lassen827', 'lassen827', 'lassen827', 'lassen827', 'lassen827', 'lassen827', 'lassen827', 'lassen827', 'lassen827', 'lassen827', 'lassen827', 'lassen827', 'lassen827', 'lassen827', 'lassen827', 'lassen827', 'lassen827']
```

\* Initially `cores_per_node` were set to `44`, but it will be changed to `40` according to the following statement:
```
Each node has two 22-core 3.45 GHz IBM POWER9 processors. Two of the
cores on each socket are reserved for system use, leaving 40 usable
cores per node.
```

This issue was encountered [here](https://github.com/radical-cybertools/radical.pilot/blob/devel/src/radical/pilot/agent/resource_manager/base.py#L172), where `cores_avail` was calculated based on env variables `LSB_HOSTS` and `LSB_MCPU_HOSTS` ([link](https://github.com/radical-cybertools/radical.pilot/blob/devel/src/radical/pilot/agent/resource_manager/lsf.py#L60)), and `self.requested_cores` is a number of cores that represents whole nodes (in my case it was 1 node, thus 44 cores) which are needed for the actual requested number of cores (I had `cores=20` in pilot description; [link](https://github.com/radical-cybertools/radical.pilot/blob/devel/src/radical/pilot/pmgr/launching/default.py#L1126))

So I understand what causes the issue, but it is not clear for me what changes should be made (kinda what logic to keep behind of these changes, especially if the case will be more complicated)

**Comment by andre-merzky** [comment]
@mtitov : can you give `lsf_summit` a try?  I think it should behave better than `lsf`.  If that's the case and we are able to use it on Lassen, we should retire the orginal `lsf` RM and switch to the new one.

**Comment by mtitov** [comment]
Issue related to this ticket was fixed - e4a931b
